src/Microsoft.AspNetCore.Routing/Matching/DfaNode.cs

```diff
@@ -30,23 +23,66 @@ public DfaNode()
         // Just for diagnostics and debugging
         public string Label { get; set; }
         
-        public List<Endpoint> Matches { get; }
+        public List<Endpoint> Matches { get; private set; }
 
-        public Dictionary<string, DfaNode> Literals { get; }
+        public Dictionary<string, DfaNode> Literals { get; private set; }
 
         public DfaNode Parameters { get; set; }
 
         public DfaNode CatchAll { get; set; }
 
         public INodeBuilderPolicy NodeBuilder { get; set; }
 
-        public Dictionary<object, DfaNode> PolicyEdges { get; }
+        public Dictionary<object, DfaNode> PolicyEdges { get; private set; }
+
+        public void AddPolicyEdge(object state, DfaNode node)
+        {
+            if (PolicyEdges == null)
+            {
+                PolicyEdges = new Dictionary<object, DfaNode>();
+            }
+
+            PolicyEdges.Add(state, node);
+        }
+
+        public void AddLiteral(string literal, DfaNode node)
+        {
+            if (Literals == null)
+            {
+                Literals = new Dictionary<string, DfaNode>(StringComparer.OrdinalIgnoreCase);
+            }
+
+            Literals.Add(literal, node);
+        }
+
+        public void AddMatch(Endpoint endpoint)
+        {
+            if (Matches == null)
+            {
+                Matches = new List<Endpoint>();
+            }
+
+            Matches.Add(endpoint);
+        }
+
+        public void AddMatches(IEnumerable<Endpoint> endpoints)
```

Comment:
Nit: 
Would it be better to use **IReadOnlyList&lt;EndPoint&gt;** instead of **IEnumerable&lt;EndPoint&gt;**?
That way if ever `List<T>.AddRange(IEnumerable<T>)` specializes for **IReadOnlyList&lt;T&gt;** we can reap the benefits of this.

Reply:
Hmm, I don't think that would happen.

Reply:
I agree it's not likely. Ignore my remark.